### PR TITLE
feat(server): initial implementation of PIN reset functionality

### PR DESCRIPTION
**Pull request by jdvlio**
### Overview

A wallet user that has forgotten their PIN should be allowed to reset it after answering one or more security questions.  These security questions and their corresponding answers are set upon account creation.  The enclave code authenticates all such reset requests using the valid responses, which were sealed and stored upon account initialization.

### Task Breakdown

The following functionality should be implemented:
- [x] A new type to facilitate the creation, ~modification~ and validation of security responses
- [x] Valid security responses should be sealed as part of the `WalletStorable` struct
- [x] Policy under which PIN resets should be authorized
- [ ] ~User security questions should be retrievable as plaintext via a sensible mechanism~
- [ ] Update backend integration tests
- [ ] Connect with frontend

### Client Functionality Required

The frontend should define the security questions users may choose from.  The backend will allow arbitrary security questions which will be retrievable via HTTP during the reset process.

The functionality in this PR requires two separate exchanges, via MessagePack, between the client and server:

#### Initial request

Name: **StartPinReset**
Request:

| Field | Type | Encoding | Description |
| - | - | - | - |
| wallet_id | string | UTF-8 | User's wallet id. | 
|  wallet_auth_map | Map<string,string> | UTF-8 (keys and values) | Security questions and attempts. |
|  client_pk| uint8[32] | - | Client public key for a newly created session. |

Response variant A:

| Field | Type | Encoding | Description |
| - | - | - | - |
| ServerPk | uint8[32] | ~base64~ | Enclave public key for this session. |

Response variant B:

| Field | Type | Encoding | Description |
| - | - | - | - |
| InvalidAuth | null | - | Invalid authentication details received. |

Response variant C:

| Field | Type | Encoding | Description |
| - | - | - | - |
| Failed | string | UTF-8 | Error message: unexpected error encountered |

#### Follow-up request

Name: **PinReset**
Request:

| Field | Type | Encoding | Description |
| - | - | - | - |
| wallet_id | string | UTF-8 | User's wallet id. | 
| new_pin | string | UTF-8 | User's new PIN. |
| new_pin_mac | uint8[32] | ~base64~| MAC of the new PIN. | 
|  client_pk| uint8[32] | - | Client public key for a newly created session. |

Response variant A:

| Field | Type | Encoding | Description |
| - | - | - | - |
| Reset | null | - | User PIN succesfully reset. |

Response variant A:

| Field | Type | Encoding | Description |
| - | - | - | - |
| InvalidAuth | null | - | Unable to authenticate MAC |

Response variant B:

| Field | Type | Encoding | Description |
| - | - | - | - |
| Failed | string | UTF-8 | Error message: failed to reset PIN. |

### Deferred functionality

The following functionality will be addressed in one or more follow-up PRs:
- Modification of saved responses to security challenges
- Plaintext retrieval of arbitrary security questions from the server

### CODE REVIEW

File: web-server/sgx-wallet-impl/src/schema/actions.rs [hunk 1]
```diff
@@ -57,6 +61,46 @@ impl From<UnlockWalletError> for OpenWalletResult {
...
 }
 
+#[derive(Clone, Eq, PartialEq, Debug)] // core
+#[derive(Deserialize, Serialize)] // serde
+#[derive(Zeroize, ZeroizeOnDrop)] // zeroize
+pub struct PinReset {
+    pub wallet_id: WalletId,
+    #[zeroize(skip)]
+    // Interior mutability pattern: needed to obtain a draining iterator
+    pub auth_attempts: RefCell<WalletAuthMap>,
```
Comment (billguo99) [inline web-server/sgx-wallet-impl/src/schema/actions.rs:71]: Can you explain why we need to obtain a draining iterator here or use "RefCell"?

File: web-server/sgx-wallet-impl/src/schema/actions.rs [hunk 2]
```diff
@@ -20,6 +22,8 @@ use crate::wallet_operations::store::UnlockWalletError;
 pub struct CreateWallet {
     pub owner_name: String,
     pub auth_pin: WalletPin,
+    #[zeroize(skip)]
+    pub challenge_map: WalletAuthMap,
```
Comment (billguo99) [inline web-server/sgx-wallet-impl/src/schema/actions.rs:26]: Can we rename this to `auth_map`?

File: web-server/sgx-wallet-impl/src/schema/actions.rs (hunk 1 above)
Comment (jdvlio) [reply web-server/sgx-wallet-impl/src/schema/actions.rs:75]: Does the latest iteration suffice?

Comment (billguo99) [reply web-server/sgx-wallet-impl/src/schema/actions.rs:75]: Yes

File: web-server/sgx-wallet-impl/src/schema/auth.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                }
+            } else {
+                return Err(AuthError::InvalidAttempt);
+            }
+        }
+
+        /*
+         *  Authentication fails iff two or more user responses were incorect
+         */
+        if mismatch_count > 1 {
```
Comment (billguo99) [inline web-server/sgx-wallet-impl/src/schema/auth.rs:85]: Should we change this to `mismatch_count > 0`? 

I thought we allow Pin Reset only when user gets all security questions correct.

File: web-server/sgx-wallet-test/app/codegen/Enclave_u.c [hunk 4]
```diff
@@ -209,6 +209,14 @@ typedef struct ms_u_open64_ocall_t {
 	int ms_mode;
 } ms_u_open64_ocall_t;
 
+typedef struct ms_u_openat_ocall_t {
```
Comment (billguo99) [inline web-server/sgx-wallet-test/app/codegen/Enclave_u.c:212]: Why are these structs getting added to `Encalve_u.c`?

**Review by billguo99 — CHANGES_REQUESTED** [review]
I'm failing to run `make` in `sgx-wallet-test`, I think I'm using older version of SGX toolchain:
```
error: linking with `cc` failed: exit status: 1
...
  = note: /usr/bin/ld: ../build/lib/libEnclave_u.a(Enclave_u.o): in function `Enclave_u_fdopendir_ocall':
          Enclave_u.c:(.text+0x3f0): undefined reference to `u_fdopendir_ocall'
          /usr/bin/ld: ../build/lib/libEnclave_u.a(Enclave_u.o): in function `Enclave_u_unlinkat_ocall':
          Enclave_u.c:(.text+0x537): undefined reference to `u_unlinkat_ocall'
          /usr/bin/ld: ../build/lib/libEnclave_u.a(Enclave_u.o): in function `Enclave_u_openat_ocall':
          Enclave_u.c:(.text+0x7a7): undefined reference to `u_openat_ocall'
          collect2: error: ld returned 1 exit status
```

Did you update your Rust SGX SDK or Intel SGX SDK toolchain?

**Comment by billguo99** [comment]
Also need to have a look at why the Github Actions checks are failing for web-server.

File: web-server/sgx-wallet-impl/src/schema/auth.rs (hunk 3 above)
Comment (jdvlio) [reply web-server/sgx-wallet-impl/src/schema/auth.rs:85]: I'm not sure.  I was under the impression that they are allowed to get at least one wrong.

Comment (billguo99) [reply web-server/sgx-wallet-impl/src/schema/auth.rs:85]: Cool 2 out of 3 correct

**Comment by billguo99** [comment]
Still getting errors when running `sgx-wallet-test` with `make`:
```
error[E0063]: missing field `auth_map` in initializer of `sgx_wallet_impl::schema::actions::CreateWallet`
  --> src/helpers/wallet_store.rs:20:20
   |
20 |     let request = &actions::CreateWallet {
   |                    ^^^^^^^^^^^^^^^^^^^^^ missing `auth_map`

error[E0063]: missing field `auth_map` in initializer of `sgx_wallet_impl::schema::actions::CreateWallet`
  --> src/wallet_operations/test_create_wallet.rs:10:20
   |
10 |     let request = &actions::CreateWallet {
   |                    ^^^^^^^^^^^^^^^^^^^^^ missing `auth_map`

For more information about this error, try `rustc --explain E0063`.
error: could not compile `sgx-wallet-test-enclave` due to 2 previous errors
```

File: web-server/sgx-wallet-test/app/codegen/Enclave_u.c (hunk 4 above)
Comment (jdvlio) [reply web-server/sgx-wallet-test/app/codegen/Enclave_u.c]: No idea.  It gets generated automatically by `bindgen`

**Comment by jdvlio** [comment]
> Still getting errors when running `sgx-wallet-test` with `make`:
> 
> ```
> error[E0063]: missing field `auth_map` in initializer of `sgx_wallet_impl::schema::actions::CreateWallet`
>   --> src/helpers/wallet_store.rs:20:20
>    |
> 20 |     let request = &actions::CreateWallet {
>    |                    ^^^^^^^^^^^^^^^^^^^^^ missing `auth_map`
> 
> error[E0063]: missing field `auth_map` in initializer of `sgx_wallet_impl::schema::actions::CreateWallet`
>   --> src/wallet_operations/test_create_wallet.rs:10:20
>    |
> 10 |     let request = &actions::CreateWallet {
>    |                    ^^^^^^^^^^^^^^^^^^^^^ missing `auth_map`
> 
> For more information about this error, try `rustc --explain E0063`.
> error: could not compile `sgx-wallet-test-enclave` due to 2 previous errors
> ```

This should be resolved now

**Review by billguo99 — APPROVED** [review]

**Pull request merged**